Grundsteinlegung Git Ninja Dojo

cheyder opened the pull request:
There are no common git conventions among the colleagues at my company. The git commit history in most of the repositories is not helpful in any way.

To change that, this repository shall contain all conventions and commands you need to maintain a sophisticated commit history. This is a first overview of all necessary conventions and commands and exercises I intend to provide.

Consider it as a request for comments

cheyder committed:
Propose all conventions, commands and excercises

There are no common git conventions among the colleagues at my company.
The git commit history in most of the repositories is not helpful in
any way.

To change that, this repository shall contain all conventions and
commands you need to maintain a sophisticated commit history. This is
a first overview of all necessary conventions and commands and
exercises I intend to provide.

Consider it as a request for comments

cheyder committed:
Provide more insights in the practical background of the conventions

The first draft is focused on the pure conventions and hard facts, but
does not provide many insights into the background of those. Practical
examples, real world experiences and the abstract conventions are now
more intertwined with each other.

As the length grow with more examples, the files are moved to dedicated
folders.

christianhaberland commented:
> nicht länger als 50 und auf keinen Fall länger als 72 Zeichen ist

Irgendwie scheint hier ein Fehler im Satz zu sein.

christianhaberland commented:
> haben einen Body, der > - nie längera ls 72 Zeichen ist

Hier scheint auch noch ein kleiner Fehler enthalten zu sein. Ich kann mir vorstellen, dass die ausführlichen Beschreibungen länger als 72 Zeichen sein soll.

dotdash commented on README.md:
```diff
@@ -0,0 +1,13 @@
+
+Dies ist ein Ort, an dem wir fortgeschrittene Fertigkeiten für die Versionierung mit Git sammeln, verbreiten und trainieren.
+
+Für alle, die sich eine Git History mit wenig Noise und dafür gehaltvollen Informationen wünschen. Für alle, die aus der Git History gerne die fachlichen und technischen Gründe für die Entstehung einer Codestelle rauslesen würden. Für alle, die große PRs gerne in kleinen, sinnvollen, zusammenhängenden Einheiten durchsehen möchten.
```
Als derjenige der das anwendet bin ich nur in zweiter Linie Nutznießer davon, als allererstes profitieren Reviewer und andere Leser (zu denen ich später ggf. auch mal gehöre). 

Maybe: Für alle, die ihren Reviewern das Leben durch kleine, eigenständige und wohl dokumentierte Commits erleichtern wollen. Für alle, die fachliche und technische Entscheidungen für andere möglichst einfach nachvollziehbar machen wollen. Für alle, die die Suche nach dem Warum in der git History von unnötigem Rauschen befreien wollen. Und für alle, die sich das auch von ihren Mitstreitern erhoffen.

cheyder replied:
Interessanter Punkt! Ich bin wohl davon ausgegangen, dass der (spätere) Eigennutz überzeugender ist, als der (unmittelbar wirksame) Altruismus. Aber ja, so rum wird vielleicht die Haltung klarer, die es erfordert: Ich mache mir etwas mehr Aufwand, auch wenn ich nicht weiß, ob ich selbst etwas davon haben werde.

dotdash commented on code_of_honour/10_commits.md:
```diff
@@ -0,0 +1,70 @@
+
+Viel hilfreicher wäre ein einzelner Commit, in dem alles enthalten ist. Wenn ich mir die Changes dieses Commits anzeigen lasse, sehe ich auf einen Blick alle Stellen, die mit dieser Änderung/Stelle zusammenhängen. Ich kann ohne weiteres nachvollziehen, warum der Code an dieser Stelle so ist, wie er ist und wo evtl. Side-effects auftreten könnten, wenn ich hier was ändere.
+
+Umgekehrt gilt natürlich das Gleiche: Ich öffne die Git-History für eine bestimmte Stelle im Code und lande bei einem Commit in der Art von `Simply everything which came to my mind while working on this poorly specified ticket I started working on two weeks ago – have fun and don't panic`. Jede Anpassung die für sich lauffähig und sinnvoll ist, kommt in einen eigenen Commit.
```
Das deckt hier zwei Aspekte ab: vollständig / minimal. Würde ich aufteilen.

```suggestion
## vollständig

> [!IMPORTANT]
> Änderungen sind mit all ihren Abhängigkeiten zusammen in einem Commit und nicht über mehrere Commits verteilt.

Gerade wenn ich einen größeren PR erstelle, kann ich meinem Reviewer helfen nicht von der Komplexität erschlagen zu werden, in dem ich meine Änderungen in einzeln reviewbare Commits aufteile. Das funktioniert nur, wenn alle Bestandteile der Änderung im jeweiligen Commit vorhanden sind. Wenn ich zum Beispiel nur ein Interface erweitere, aber die neue Methode nicht verwende und/oder nicht in den entsprechenden Klassen implementiere, ist mein Reviewer oder ein späterer Leser des Codes darauf angewiesen sich die Einzelteile zusammen zu suchen, was mehr Aufwand bedeutet und die Gefahr birgt etwas zu übersehen.

Weitere Beispiele für Dinge die in der Regel zusammengehören:
 * Migrationen und Modellanpassungen
 * Bugfix + Regressionstest
 * Neues Feature + Tests
 * Funktionsanpassungen + Tests
 
## minimal

> [!IMPORTANT]
> Änderungen, die nicht zwingend zusammengehören fließen in separate Commits.

Das Gegenstück zur Vollständigkeit ist die Minimalität eines Commits. Jede Änderung die unnötig mit in einen Commit aufgenommen wird, erhöht dessen Komplexität. Zum Einen muss mein Reviewer sich erstmal erarbeiten, welche Teile denn interagieren und welche unabhängig voneinander sind, zum Anderen resultiert dadurch auch mehr Aufwand beim Lesen der Änderungen, weil für jede Teil des Commits mal die einen und mal Teile relevant bzw. irrelevant sind und dadurch ggf. viel mehr als nötig im Diff gesprungen werden muss.

Weitere relevante Aspekte sind Tools wie `git revert` um z.B. Änderungen rückgängig zu machen. Code Style Fixes und eine fehlerhafte Änderung sind zusammen in einem Commit gelandet? `git revert` dreht beides zurück. Wären die Style Fixes einmal vorab gemacht worden, könnte man die fehlerhafte Änderung einzeln reverten, ohne den Coding Style wieder kaputt zu machen oder das ganze manuell zu machen. 
```

dotdash commented on code_of_honour/10_commits.md:
```diff
@@ -0,0 +1,70 @@
+
+## TLDR;
+
+Siehe die farbigen Boxen, am Anfang der Abschnitte.
```
```suggestion
Ich will meine Commits so strukturieren, dass ich die Arbeit für meinen Reviewer und spätere Leser des Codes (inkl. mir selbst!) vereinfache: Minimal, in sich abgeschlossen und ohne bekannte (neue) Bugs.

Siehe auch die farbigen Boxen am Anfang der Abschnitte.
```

dotdash commented on code_of_honour/10_commits.md:
```diff
@@ -0,0 +1,70 @@
+
+Ich habe ein Feature, das kaputt ist, weiß aber noch, dass es in einer früheren Version auf jeden Fall funktioniert hat. Ich möchte z.B. mit `git bisect` herausfinden, seit welchem Commit es kaputt ist. Wenn ich in der Reihe zu testender Commits, dann welche habe, die gar nicht lauffähig sind, in dem die Tests oder Checks fehlschlagen, erschwert das die Suche.
+
+Ich schaue einen großen PR Commit für Commit durch. In einem Commit sind Anpassungen, die gar nichts mit dem Rest zu tun haben. Sie sind nur dafür da, einen Fehler, der in einem vorherigen Commit eingebaut wurde, zu beheben. Dabei könnte ich mit einem interaktiven Rebase, diesen Fix direkt in den vorherigen Commit einbauen.
```
Würde ich als letzten Punkt sehen, ist eigentlich eine implizite Folge von "zusammenhängend/vollständig".

```suggestion
## lauffähig

> [!IMPORTANT]
> Alle Tests, Checks, Migrationen laufen durch.

Ist ein Teilaspekt der Vollständigkeit (s.o.). Ein Commit der nicht alle Checks besteht ist ja per Definition unvollständig, da fehlt ja ein Bugfix. Für meinen Reviewer ist es dann um so ärgerlicher, wenn ich den neuen Bug zwar behoben habe, aber erst in einem späteren Commit des gleichen PRs. Nachdem ich alles schön aufbereitet habe, damit das Review Commit für Commit erfolgen kann, sind in den einzelnen Commits dann doch Bugs drin, die der Reviewer auch identifiziert, nur um am Ende zu sehen, dass die dann im Nachgang doch noch beseitigt wurden. Häufig sieht man sowas ab der zweiten Reviewrunde, wenn die Commits nicht aktualisiert wurden, um die Bugs direkt zu beseitigen, sondern nur ein oder mehrere Commits der Form `Apply PR feedback` erstellt werden, die die Prinzipien zur Commit Struktur komplett über den Haufen werfen.
Ich habe ein Feature, das kaputt ist, weiß aber noch, dass es in einer früheren Version auf jeden Fall funktioniert hat. Ich möchte z.B. mit `git bisect` herausfinden, seit welchem Commit es kaputt ist. Wenn ich in der Reihe zu testender Commits, dann welche habe, die gar nicht lauffähig sind, in dem die Tests oder Checks fehlschlagen, erschwert das die Suche.

Ein weiterer Aspekt hier sind Tools wie `git bisect` mit denen ich bestimmen kann in welchem Commit ein Bug entstanden ist. Um so mehr Commits nicht eigenständig lauffähig sind, um so weniger exakt kann ich die Entstehung eingrenzen, und um so weniger nützlich werden solche Werkzeuge.
```

dotdash commented on code_of_honour/10_commits.md:
````diff
@@ -0,0 +1,70 @@
+> [!IMPORTANT]
+> Die Commit-Message enthält die Auswirkung der Änderung, eine Beschreibung des Zustands davor, und eine Zusammenfassung der Änderung.
+
+Das betrifft vor allem die Commit-Message. Ich schaue mal wieder in die Git-History mit der langsam sterbenden Hoffnung, dort irgendeine hilfreiche Information zu dieser absolut nicht nachvollziehbaren Stelle im Code zu finden. Aber ich finde:
+
+```
+Fix bug
+Work in progress
+Remove something
+Style
+Add it again
+Current state
+Typo
+Fix unit tests
+< the name of a class which can be found in this commit >
+< a more or less random word, associated with the changes >
+```
+
+Was mir geholfen hätte:
+
+```
+Fix <false behaviour>
+```
+
+Ah interessant! Ich weiß sofort, was dieser Commit machen soll und vor allem, ob er für mich gerade interessant sein könnte. Wenn er das könnte, kann ich für Details auch noch in den Body schauen:
+
+```
+Fix <false behaviour>
+
+Describe the false behaviour and its cause shortly. Describe shortly
+the change and how it fixes the bug.
+
+If there is a Commit, which introduced this bug, provide its ID
+If there is a link to an issue or documentation which describes the
+background of this change, add this as well
+
+<Ticket-Number>
+``` 
+
+Alright! Das erklärt mir, warum es ausgerechnet so gelöst wurde und falls ich an der Stelle was ändern muss, weiß ich auch sofort, mit welchen anderen Stellen sie zusammenhängt. Danke Developer aus der Vergangenheit, Du hast mir gerade meine Arbeit ein Stück leichter gemacht!
````
Viel zu passiv aggressiv 😝

````suggestion
Neben den impliziten Informationen darüber vermittelt hat, welche Änderungen zusammengehören, welche unabhängig voneinander sind und welche ggf. aufeinander aufbauen, die ich meinem Reviewer durch die Aufteilung in Commits vermittle, habe ich mir während der Entwicklung oft auch fachliche und technische Informationen oder Entscheidungen erarbeitet, die meinem Reviewer nützlich bis unabdingbar für ein gutes Review sind. Ein guter Platz dafür ist die Commit Message, denn die Bündelung von Änderung und deren Dokumentation hat Vorteile gegenüber Kommentaren im Code oder externen Systemen:
 * Die Änderung ist fix, also kann die Commit Message nicht veralten
 * Die Commit Message bezieht sich auf das was geändert wurde, und nichts sonst
 * Die Commit Message enthält nur genau das was für den Commit nötig ist
 * Man kann vom Code zur Commit Message finden (`git blame`, `git log`, etc.)
 * Man kann den historischen Kontext herstellen (war das schon immer falsch, oder hat sich was geändert und hier wurde eine Anpassung vergessen?)
 * Die Commit Message ist immer direkt verfügbar ohne mehrere Systeme durchforsten zu müssen

Um eine gute Commit Message zu schreiben, sollte ich mir überlegen, was ich selbst gern sehen würde, wenn ich den Commit reviewen müsste bzw. wenn ich das Commit Log durchsuchen würde um relevante Commits für ein Problem zu finden. Einige mögliche Aspekte sind:

 * Welches Problem soll gelöst werden?
 * Wie/wann äußert sich das Problem?
 * Was ist die Ursache für das Problem?
 * Wie kann man das Problem lösen?
 * Warum ist das die korrekte Lösung?
 * Welche Alternativen gibt es? Warum habe ich diese nicht genommen?
 * In welchem Bereich des Projekts tritt das Problem auf?
 * War das schon immer so, oder ist das durch einen bestimmten Commit eingeführt worden? 
 * Wenn ja, warum hat der Commit das kaputt gemacht? Fehlannahme? Was vergessen? 
 * Ticketnummer
 * Versionsnummern von Paketupdates
 * ...
 
Zum Vergleich ein Beispiel für Commit Messages die jeweils den gleichen imaginativen Commit beschreiben:

```
Fix import bug
```

vs.

```
CSV-Import: Fix crash when customer has no first name

The CSV import always replaces empty strings with null values,
and this causes a crash later on because the customer must have a
string as their first name.

Adjusting the CSV import to keep empty string would need lots of adjustments
and we must also support data sets that don't contain first names at all. So
it seems prudent to convert null values to empty strings when assigning the value
to the customer.  Note that we can't simply ignore null values, because we might
be updating an existing customer and must not keep the previous value!
```

Mit der ersten Commit Message fehlen meinem Reviewer quasi alle Informationen die er benötigt, um das Problem einzuordnen, nachzustellen oder die Lösung beurteilen zu können. In der zweiten Commit Message ist klar welcher Import betroffen ist (ggf. gibt es mehrere), mit welchen Daten das Problem ausgelöst werden kann, warum das Problem auftritt, für welche Lösung ich mich entschieden habe, warum ich dies tat, und weshalb ich eine andere mögliche Lösung, die dem Reviewer evtl. auch eingefallen wäre, verworfen habe.

Mit der Commit Message allein hat der Reviewer hier schon alle Informationen, die ich mir als Entwickler evtl. langwierig erarbeiten musste, und kann den prinzipiellen Lösungsansatz hoffentlich schon allein damit beurteilen. Im Code muss er dann "nur" noch schauen, ob diese Lösung auch wirklich umgesetzt wurde und korrekt und vollständig ist.
````

cheyder replied:
Haha, bei einem Blick in die Commit History wird doch niemand passiv aggressiv :smile: 

Ich habe den ersten Teil etwas kondensiert übernommen, bei dem zweiten Teil, der eine gute Commit Message beschreibt, befürchte ich, dass "sollte ich mir überlegen" und "einige mögliche Aspekte" als Boundaries/Orientierung zu soft sind. Außerdem lösen Ansammlungen von Fragen zumindest bei mir immer aus, dass ich sie überspringe :wink: 

Ich würde hier lieber eine Art Grundform von guten Commit Messages haben, zu der ich bei der Verwendung dieses Leitfadens immer wieder zurückkehren kann.

Hab den dazugehörigen Teil nach `code_of_honour/20_commit_messages.md` in einen eigenen Abschnitt zu Inhalt ausgelagert. Schau mal dort, ob das so passen könnte.

dotdash commented on code_of_honour/20_commit_messages.md:
```diff
@@ -0,0 +1,72 @@
+>     - Links zu Dokumentationen
+>     - Links zu Issues bei GitHub
+> - haben eine Leerzeile zwischen
+    >   - Subject und Body
```
Ohne die Leerzeile wäre es ja nur ein Subject 😂

dotdash commented on code_of_honour/20_commit_messages.md:
```diff
@@ -0,0 +1,72 @@
+>   - im Imperativ verfasst ist
+>   - das Ziel/den Effekt der Änderung beschreibt, nicht die technische Anpassung
+> - haben einen Body, der
+    >   - nie längera ls 72 Zeichen ist
```
```suggestion
> - haben einen Body, der
    >   - Fließtext in der Regel nach 72 Zeichen umbricht
```

Sonst wird's eng, wenn man z.B. ASCII Art drin haben will, oder ne Liste mal bis Spalte 110 geht oder so. Da ist das einzeilige Format dann ggf. immer noch besser als alles über einen Kamm zu scheren.

dotdash commented on code_of_honour/20_commit_messages.md:
```diff
@@ -0,0 +1,72 @@
+    >   - nicht mit einem Punkt endet
+>   - nicht länger als 50 und auf keinen Fall länger als 72 Zeichen ist
+>   - im Imperativ verfasst ist
+>   - das Ziel/den Effekt der Änderung beschreibt, nicht die technische Anpassung
```
```suggestion
>   - die Zielsetzung der Änderung zusammenfasst
>   - im Imperativ verfasst ist (`Add foo` statt `Added foo`)
>   - Möglichst prägnant ist, irgendwo < 72 Zeichen ist gut
>   - Keinen Punkt am Ende hat
```

dotdash commented on code_of_honour/20_commit_messages.md:
```diff
@@ -0,0 +1,72 @@
+> - haben eine Leerzeile zwischen
+    >   - Subject und Body
+>   - jedem Absatz im Body
+> - enthalten in der letzten Zeile die Ticketnummer
```
Nicht jeder Commit hat eine Ticketnummer.

```suggestion
> - eine eventuelle Ticketnummer in der letzten Zeile enthalten
```

Wobei ich die Verortung der Ticketnummer nicht unbedingt so starr machen würde.

dotdash commented on code_of_honour/20_commit_messages.md:
```diff
@@ -0,0 +1,72 @@
+# Commit Messages
```
Hier fehlt's an Motivation für ein paar der Konventionen.

Zum Beispiel:

Imperativ -> git Standardkonvention, siehe Default-Commit Messages von `git merge`, `git revert` etc.

Kurzer Betreff -> Betreff wird von git besonders behandelt, z.B. `git log --oneline`, `git shortlog` etc. Und das Ding dient halt der schnellen Einordnung, bei E-Mails schreib ich ja auch nicht den kompletten Inhalt in den Betreff.

Leerzeile nach dem Betreff: Durch git vorgebeben, bis zur ersten Leerzeile ist's der Betreff, danach der Body.

dotdash commented on code_of_honour/20_commit_messages.md:
````diff
@@ -0,0 +1,72 @@
+
+## 3.  Grad: Auswirkung
+```
+Receive always the same email when fetching by UID
````
Satzbau 😅

```suggestion
Always receive the same email when fetching by UID
```

Und `receive` ist passiv, sollte das `retrieve` oder `return` sein?

dotdash commented on code_of_honour/20_commit_messages.md:
````diff
@@ -0,0 +1,72 @@
+Receive always the same email when fetching by UID
+```
+
+Idealerweise sagt mir das Subject einer Commit-Message nichts darüber, was technisch geändert wurde, sondern welchen Effekt die Änderung hat, bzw. haben soll.
````
Hm, naja, und bei rein technischen Änderungen? 😅 Die Unterscheidung ist ja eher Ziel/Lösung, wobei das manchmal sogar das gleiche ist, z.B. `Remove dead method Foo::bar` 😃 

````suggestion
Idealerweise sagt mir das Subject welches Ziel durch den Commit erreicht wird ohne unötige Details zur Umsetzung. Zum Beispiel: `Beschleunige Suche im Postfach` vs. `Verwende eine Common Table Expression mit Window Functions um Transfers zu filtern`.
````

dotdash commented on code_of_honour/20_commit_messages.md:
````diff
@@ -0,0 +1,72 @@
+
+## 4. Grad: Body
+```
+Receive always the same email when fetching by UID
````
```suggestion
Always receive the same email when fetching by UID
```

dotdash commented on code_of_honour/20_commit_messages.md:
````diff
@@ -0,0 +1,72 @@
+```
+Receive always the same email when fetching by UID
+
+Sometimes we try to fetch an email via imap from the users mailbox and
+don't receive the correct one, but another one. The reason is that
+fetching an email with its UID is error prone as the UID only MUST not
+change during a session, but may change between session, which MUST be
+detectable by comparing the UIDVALIDITY.
+
+To prevent that, the UID is always used/persisted with its UIDVALIDITY
+and when fetching the next time by UID, the corresponding UIDVALIDITY
+is checked against the current one on server side. If they differ, the
+UID has to be fetched again.
````
Hm, vllt. nicht das beste Beispiel 😅 Mir ist zumindest absolut nicht klar, wie das Problem da gelöst wird. Selbst wenn du dir jetzt UIDVALIDITY und UID merkst, hast du bei einer Änderung des serverseitigen UIDVALIDITY Wertes ja trotzdem nur noch 2 Werte die keine Gültigkeit mehr besitzen.

cheyder replied:
`the UID has to be fetched again` – es dient nur der Invalidierung der UID, alles was danach passiert war nicht Bestandteil des Commits. Aber wir müssen ja hier keine alten Kamellen aufwärmen. Ich nehm einfach Deine Beispiel Commit Message :wink:

dotdash commented on code_of_honour/30_branching.md:
```diff
@@ -0,0 +1,7 @@
+# Branching
```
```suggestion
# Rebasing
```
?

Das hat mit Branching ja nicht viel zu tun.

cheyder replied:
Hmm... ja, mir gehts halt eigentlich darum, wie wir uns "auf den Branches bewegen". Vielleicht "Commit History"?

dotdash commented on techniques/20_rearrange.md:
```diff
@@ -0,0 +1,15 @@
+| Command                                            |                                                                                                         |
+|----------------------------------------------------|---------------------------------------------------------------------------------------------------------|
+| `git stash`                                        | Alle nicht committeten Änderungen beiseite legen                                                        |
+| `git stash pop`                                    | Den letzten Eintrag im Stash hervorholen                                                                |
```
```suggestion
| `git stash pop`                                    | Den letzten Eintrag im Stash anwenden und entfernen                                                                |
```

dotdash commented on techniques/20_rearrange.md:
```diff
@@ -0,0 +1,15 @@
+| `git stash list`                                   | Alle Einträge im Stash sehen                                                                            |
+| `git stash --keep-index`                           | Alle nicht committeten Änderungen beiseite legen, aber Staging Area behalten                            |
+| `git commit --amend`                               | Änderung in der Staging Area zum letzten Commit hinzufügen                                              |
+| `git rebase master`                                | Den Unterschied zwischen dem aktuellen Branch und master, an master dranhängen.                         |
```
"an master dranhängen" klingt als würde `master` geändert

```suggestion
| `git rebase master`                                | Den aktuellen Branch inkl. History auf master basierend neu aufbauen                         |
```

dotdash commented on techniques/20_rearrange.md:
```diff
@@ -0,0 +1,15 @@
+| `git stash --keep-index`                           | Alle nicht committeten Änderungen beiseite legen, aber Staging Area behalten                            |
+| `git commit --amend`                               | Änderung in der Staging Area zum letzten Commit hinzufügen                                              |
+| `git rebase master`                                | Den Unterschied zwischen dem aktuellen Branch und master, an master dranhängen.                         |
+| `git rebase --onto <Branch> HEAD~<Anzahl Commits>` | Die letzten <Anzahl Commits> auf dem aktuellen Branch nehmen und an <Branch> dranhängen.                |
```
Das ist inhaltlich falsch sobald da Merge Commits involviert sind. Die Anzahl passt nur entlang der First Parent Historie. Siehe oben bzgl. "dranhängen"

dotdash commented on techniques/20_rearrange.md:
```diff
@@ -0,0 +1,15 @@
+| `git commit --amend`                               | Änderung in der Staging Area zum letzten Commit hinzufügen                                              |
+| `git rebase master`                                | Den Unterschied zwischen dem aktuellen Branch und master, an master dranhängen.                         |
+| `git rebase --onto <Branch> HEAD~<Anzahl Commits>` | Die letzten <Anzahl Commits> auf dem aktuellen Branch nehmen und an <Branch> dranhängen.                |
+| `git rebase -i HEAD~<Anzahl Commits>`              | Interaktives Rebase der letzten <Anzahl Commits> auf dem aktuellen Branch.                              |
```
Siehe oben bzgl. "Anzahl". Maybe:

```suggestion
| `git rebase -i HEAD~<n>`              | Interaktiver Rebase des aktuellen Branches ab Commit `HEAD~n`                              |
```

dotdash commented on techniques/20_rearrange.md:
```diff
@@ -0,0 +1,15 @@
+| `git rebase --onto <Branch> HEAD~<Anzahl Commits>` | Die letzten <Anzahl Commits> auf dem aktuellen Branch nehmen und an <Branch> dranhängen.                |
+| `git rebase -i HEAD~<Anzahl Commits>`              | Interaktives Rebase der letzten <Anzahl Commits> auf dem aktuellen Branch.                              |
+| `git push --force-with-lease`                      | Erneutes pushen nach einem Rebase, ohne mögliche Änderungen im Repo zu überschreiben.                   |
+| `git pull --rebase`                                | Den Remote-Branch pulln, nachdem er gerebased wurde.                                                    |
```
Nee, das ist komplett daneben. Das macht ein fetch+rebase, sprich der ausgecheckte Branch wird dadurch rebased. Mit "nachdem rebase" hat das gar nichts zu tun.

cheyder replied:
Es geht mir hier darum, dass ich von einem feature branch (FB1) abzweige (FB2) und von dort weiter entwickle. Währenddessen wird FB1 gerebased und gepusht. Um diese Änderungen, bzw. den aktuellen FB1 in meinen Branch (FB2) zu kriegen, nutze ich immer `git pull --rebase`. Deshalb steht da "nachdem er (der remote brach) gerebased wurde".

Fändest Du hier eine kurze Beschreibung des Szenarios besser?

|||
|---|---|
|`git pull --rebase`| Wenn von einem anderen Branch (B1) abgezweigt wurde (B2) und B1 danach gerebased werden sollte, lassen sich die Änderungen aus B2 mit diesem Befehl auf den aktuellen  B1 neu aufbauen|

dotdash commented on techniques/20_rearrange.md:
```diff
@@ -0,0 +1,15 @@
+| `git rebase -i HEAD~<Anzahl Commits>`              | Interaktives Rebase der letzten <Anzahl Commits> auf dem aktuellen Branch.                              |
+| `git push --force-with-lease`                      | Erneutes pushen nach einem Rebase, ohne mögliche Änderungen im Repo zu überschreiben.                   |
+| `git pull --rebase`                                | Den Remote-Branch pulln, nachdem er gerebased wurde.                                                    |
+| `composer update --lock`                           | Nur den Hash in `composer.lock` updaten, wenn unterschiedliche `composer.json` Dateien gemerged werden. |
```
Was hat Composer hier zu suchen? Und je nachdem wie die konkreten Änderungen aussehen ist das auch nicht die richtige Lösung für den Konflikt.

cheyder replied:
Ist notwendiger Bestandteil in meinem Werkzeugkasten für interaktives Rebasing und deshalb hier.
Was hältsts du von folgender Anpassung?

|||
|-|-|
|`composer update --lock`| Sollten während eines Rebase zwei  `composer.json` gemerged worden sein und dabei infolge unterschiedlicher Änderungen an den Paketen der Hash in `composer.lock` divergieren, lässt sich mit diesem Befehl nur dieser Hash updaten |

dotdash commented on techniques/20_rearrange.md:
```diff
@@ -0,0 +1,15 @@
+| `git stash pop`                                    | Den letzten Eintrag im Stash hervorholen                                                                |
+| `git stash list`                                   | Alle Einträge im Stash sehen                                                                            |
+| `git stash --keep-index`                           | Alle nicht committeten Änderungen beiseite legen, aber Staging Area behalten                            |
+| `git commit --amend`                               | Änderung in der Staging Area zum letzten Commit hinzufügen                                              |
```
`git commit --fixup=<commit-id>` ist vllt. nach interessant, das macht einen fixup Commit für ein späteres Rebase fertig.

dotdash commented on techniques/30_reset.md:
```diff
@@ -0,0 +1,8 @@
+| Command                      |                                                                            |
+|------------------------------|----------------------------------------------------------------------------|
+| `git reflog`                 | Der Blick ins Git des Git                                                  |
+| `git reset --soft <Commit>`  | Gehe zurück zu <Commit>, behalte alles, dort wo es war.                    |
```
Das "behalte alles, dort wo es war" ergibt speziell mit der Beschreibung von `--mixed` gar keinen Sinn.

```suggestion
| `git reset --soft <Commit>`  | Setze HEAD auf <Commit>, ohne Index und Working Tree zu verändern                    |
```

cheyder replied:
Es ging bei diesen Formulierungen darum, diese Befehle Leuten zugänglich zu machen, die noch keine Zeit hatten sich das imho extrem schwer zugängliche Git Grundvokabular anzusehen.

Wenn ich weiß dass bei soft sich nichts ändert, außer, dass ich wieder auf einem anderen Commit bin und bei mixed am Ende alle Änderungen noch da sind, aber halt im Working Directory liegen, kann ich diese Befehle immerhin schon mal halbwegs treffsicher verwenden. 

Wenn das für dich noch keinen Sinn ergab, wird es damit hoffentlich klarer. Es war halt ein Abwägen von Verständlichkeit/Zugänglichkeit, Korrektheit/Genauigkeit und Nützlichkeit. Bei deiner Beschreibung würde ich z.B. auch, wenn mich nicht alles täuscht, jedes Mal in einem detached head state landen.

Ich habe nichtsdestotrotz nochmal eine Sektion mit Git Basics und Vokabular eingeführt, damit an solchen stellen auch technisch korrekte und präzise Formurlierungen verwendet werden können. Ich überlege trotzdem sie um eine "auf Deutsch gesagt" Spalte zu ergänzen. Siehst du ja dann in den Änderungen

dotdash replied:
> und bei mixed am Ende alle Änderungen noch da sind

Sind sie eben nicht, alles was (nur) im Index war wird ja weggeworfen:

```
derp $ git init
Initialized empty Git repository in /home/bsteinbrink/src/derp/.git/

derp (master) $ touch foo
derp (master) $ git add foo
derp (master) $ git commit -m init
[master (root-commit) fd7fdd6] init
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 foo

derp (master) $ echo a > foo
derp (master) $ git add foo
derp (master) $ echo b > foo

derp (master) $ git status
On branch master
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   foo

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   foo

derp (master) $ git diff --cached
diff --git foo foo
index e69de29..7898192 100644
--- foo
+++ foo
@@ -0,0 +1 @@
+a

derp (master) $ git diff
diff --git foo foo
index 7898192..6178079 100644
--- foo
+++ foo
@@ -1 +1 @@
-a
+b


derp (master) $ git reset --mixed HEAD
Unstaged changes after reset:
M	foo
derp (master) $ git diff --cached
derp (master) $ git diff
diff --git foo foo
index e69de29..6178079 100644
--- foo
+++ foo
@@ -0,0 +1 @@
+b
```

Der Zwischenstand den ich im Index / der Staging Area hatte, bei dem `a` eingefuegt wurde ist futscht.

> Genauso ist es für die Verwendung der Befehle zweitrangig, dass ich bei deiner Beschreibung jedes Mal in einem detached head state landen würde.

Reset biegt ja den ausgecheckten Branch Head um, HEAD bleibt an diesen Branch Head attached. Fuer nen Detached HEAD braucht's nen Checkout.

cheyder replied:
Bin endlich mal dazu gekommen, es für mich nochmal nachzustellen um es auseinander zu friemeln. Die gestagete Änderung wird doch eher zurück ins Working Directory geschoben, oder? Futsch ist sie ja bei Dir nur, weil sie im Working Directory bereits überschrieben wurde. Also zumindest war es bei meinem Nachstellen so, dass die bereits gestagete Änderung nicht einfach verschwand, sondern eher unstaged wurde.

```
cheyder@NB000606:~/dev/git_test$ git diff --staged
diff --git a/test b/test
index b1e6722..8422d40 100644
--- a/test
+++ b/test
@@ -1,3 +1,4 @@
 A
 B
 C
+D

cheyder@NB000606:~/dev/git_test$ git diff
diff --git a/test b/test
index 8422d40..8fda00d 100644
--- a/test
+++ b/test
@@ -2,3 +2,4 @@ A
 B
 C
 D
+E

cheyder@NB000606:~/dev/git_test$ git reset --mixed HEAD
Nicht zum Commit vorgemerkte Änderungen nach Zurücksetzung:
M	test

cheyder@NB000606:~/dev/git_test$ git diff --staged

cheyder@NB000606:~/dev/git_test$ git diff
diff --git a/test b/test
index b1e6722..8fda00d 100644
--- a/test
+++ b/test
@@ -1,3 +1,5 @@
 A
 B
 C
+D
+E
```

Habe es auch nochmal für `--soft` getestet und da das Verhalten gehabt, dass alles was commitet oder gestaged war, im Index gelandet ist und alles was im Working Directory war dort blieb. Was meiner starken Vereinfachung von "behalte alles dort wo es war" entspricht, aber vielleicht wirklich zu schwammig war. 

Habe die Formulierungen jetzt nochmal entsprechend angepasst. Falls du Lust hast, kannst du ja nochmal drüber schauen.

dotdash commented on techniques/30_reset.md:
```diff
@@ -0,0 +1,8 @@
+|------------------------------|----------------------------------------------------------------------------|
+| `git reflog`                 | Der Blick ins Git des Git                                                  |
+| `git reset --soft <Commit>`  | Gehe zurück zu <Commit>, behalte alles, dort wo es war.                    |
+| `git reset --mixed <Commit>` | Gehe zurück zu <Commit>, uncommitte und unstage alle Änderungen bis dahin. |
```
```suggestion
| `git reset --mixed <Commit>` | Setze HEAD auf <Commit> und setze den Index zurück, ohne den Working Tree zu verändern --> keine gestagten Änderungen mehr |
```

dotdash commented on techniques/30_reset.md:
```diff
@@ -0,0 +1,8 @@
+| `git reflog`                 | Der Blick ins Git des Git                                                  |
+| `git reset --soft <Commit>`  | Gehe zurück zu <Commit>, behalte alles, dort wo es war.                    |
+| `git reset --mixed <Commit>` | Gehe zurück zu <Commit>, uncommitte und unstage alle Änderungen bis dahin. |
+| `git reset --hard <Commit>`  | Gehe zurück zu <Commit>, verwerfe alles bis dahin.                         |
```
```suggestion
| `git reset --hard <Commit>`  | Setze HEAD auf <Commit>, setze Index und Working Tree zurück --> Keine uncomitteten Änderungen mehr |
```

cheyder committed:
Introduce basic git concepts and vocabulary

To be able to use more precise formulations later on, an introduction in
the git basics of pointers (branches, commits, HEAD) and navigation with
the help of suffixes is provided.

cheyder committed:
Rename the repository to Git Dojo

It seemed to me more handy to use the short version

cheyder committed:
Introduce the "minimal backlog"

cheyder committed:
Add small overview of states and refine formulation of reset commands
